frontend/app/hooks/useCards.ts

```diff
@@ -0,0 +1,188 @@
+import {
+  useMutation,
+  useQueryClient,
+  useInfiniteQuery,
+} from '@tanstack/react-query';
+import {
+  fetchCards,
+  fetchFavoriteCards,
+  updateIdiom,
+  updateIdiomVote,
+  CARDS_PER_PAGE,
+} from '../services/cardService';
+import { CardData } from '../types/card';
+
+interface InfiniteQueryData {
+  pages: CardData[][];
+  pageParams: unknown[];
+}
+
+export const queryKeys = {
+  cards: ['cards'] as const,
+  cardsWithSearch: (search?: string) => ['cards', search] as const,
+  favorites: ['favorites'] as const,
+  card: (id: string) => ['card', id] as const,
+};
+
+const useInfiniteCardQuery = (
+  queryKey: readonly unknown[],
+  queryFn: (pageParam: number) => Promise<CardData[]>,
+) => {
+  return useInfiniteQuery({
+    queryKey,
+    queryFn: ({ pageParam = 1 }) => queryFn(pageParam),
+    getNextPageParam: (lastPage, allPages) => {
+      return lastPage.length === CARDS_PER_PAGE
+        ? allPages.length + 1
+        : undefined;
+    },
+    initialPageParam: 1,
+    enabled: true,
+  });
+};
+
+export const useCards = (search?: string) => {
+  return useInfiniteCardQuery(queryKeys.cardsWithSearch(search), (pageParam) =>
+    fetchCards(pageParam, CARDS_PER_PAGE, search),
+  );
+};
+
+export const useFavoriteCards = () => {
+  return useInfiniteCardQuery(queryKeys.favorites, (pageParam) =>
+    fetchFavoriteCards(pageParam, CARDS_PER_PAGE),
+  );
+};
+
+const updateCacheData = (
+  queryClient: ReturnType<typeof useQueryClient>,
+  queryKey: readonly unknown[],
+  updateFn: (card: CardData) => CardData,
+) => {
+  queryClient.setQueriesData(
+    { queryKey },
+    (oldData: InfiniteQueryData | undefined) => {
+      if (!oldData) return oldData;
+
+      return {
+        ...oldData,
+        pages: oldData.pages.map((page: CardData[]) => page.map(updateFn)),
+      };
+    },
+  );
+};
+
+export const useUpdateFavorite = () => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: ({ cardId, favorite }: { cardId: string; favorite: boolean }) =>
+      updateIdiom(cardId, favorite),
+
+    onMutate: async ({ cardId, favorite }) => {
```

Comment:
This looks really complicated, what is this doing? Should we leave a comment for future us here? Let's make sure we understand what this does, otherwise changing it in the future will be a pain